src/dataset.rs

```diff
@@ -165,18 +165,31 @@ pub fn filter_outlier(
         // remove 5% of items from each sub group
         let total = sub_groups.iter().map(|(_, vec)| vec.len()).sum::<usize>();
         let mut has_been_removed = 0;
+        let mut to_rm_index = HashSet::new();
 
         for (delta_t, sub_group) in sub_groups.iter().rev() {
             if has_been_removed + sub_group.len() > total / 20 {
                 filtered_items.extend_from_slice(sub_group);
             } else {
                 has_been_removed += sub_group.len();
-                // keep the items in trainset if they are not removed from filtered_items
-                trainset.retain(|item| {
-                    item.reviews[0].rating != rating || item.reviews[1].delta_t != *delta_t
-                });
+                to_rm_index.extend(
+                    trainset
```

Comment:
You're iterating over the trainset in a loop, so this is still O(n*m). Can't you put (rating, delta_t) in the hashset directly, then do a single retain at the end that checks each item to see if it's in the hashset?